### PR TITLE
interfaces: add new upower-control interface

### PR DESCRIPTION
**Pull request by morphis**

### CODE REVIEW

**Comment by morphis** [comment]
@jdstrand you have time to review and approve from the security side?

File: interfaces/builtin/upower_control.go [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+sendmsg
+sendto
+recvfrom
+`
+
+type UpowerControlInterface struct{}
+
+func (iface *UpowerControlInterface) Name() string {
+	return "upower-control"
+}
```
Comment (jdstrand) [inline interfaces/builtin/upower_control.go]: Please see my comments in the upower-observe PR: https://github.com/snapcore/snapd/pull/2469#discussion_r95171985

File: interfaces/builtin/upower_control.go [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+dbus (receive, send)
+    bus=system
+    path=/org/freedesktop/UPower{,/**}
+    interface=org.freedesktop.DBus.*,
+
+dbus (receive, send)
+    bus=system
+    path=/org/freedesktop/login1{,/**}
+    interface=org.freedesktop.DBus.*
+    peer=(label=unconfined),
```
Comment (jdstrand) [inline interfaces/builtin/upower_control.go]: Is receive actually needed here?

File: interfaces/builtin/upower_control.go [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+# Allow access to logind service as we need to query it for possible
+# power states and trigger these when the battery gets low and the
+# system enters a critical state.
+dbus (send)
+    bus=system
+    path=/org/freedesktop/login1{,/**}
+    interface=org.freedesktop.login1.Manager
+    member={CanPowerOff,CanSuspend,CanHibernate,CanHybridSleep,PowerOff,Suspend,Hibernate,HybrisSleep},
+    peer=(label=unconfined),
```
Comment (jdstrand) [inline interfaces/builtin/upower_control.go]: The previous rule makes this rule redundant. Should there only be this one rule?

File: interfaces/builtin/upower_control.go [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+`
+
+const upowerControlPermanentSlotSeccomp = `
+bind
+getsockname
```
Comment (jdstrand) [inline interfaces/builtin/upower_control.go]: getsockname is in the default seccomp template.

File: interfaces/builtin/upower_control.go [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+  <allow send_destination="org.freedesktop.UPower"
+         send_interface="org.freedesktop.UPower"/>
+  <allow send_destination="org.freedesktop.UPower"
+         send_interface="org.freedesktop.UPower.Device"/>
+  <allow send_destination="org.freedesktop.UPower"
+         send_interface="org.freedesktop.UPower.KbdBacklight"/>
+  <allow send_destination="org.freedesktop.UPower"
+         send_interface="org.freedesktop.UPower.Wakeups"/>
+</policy>
```
Comment (jdstrand) [inline interfaces/builtin/upower_control.go]: Are all of these considered safe for everyone or just root? If everyone, then perhaps that is an argument for merging upower-control into upower-observe....

File: interfaces/builtin/upower_control.go [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+sendto
+recvfrom
+`
+
+const upowerControlPermanentSlotDBus = `
+<!-- DBus policy for upower (upstream version 0.99.4) -->
+<policy user="root">
+  <allow own="org.freedesktop.UPower"/>
+</policy>
+<policy context="default">
```
Comment (jdstrand) [inline interfaces/builtin/upower_control.go]: Can you add:

```
<deny own="org.freedesktop.UPower"/>
```

File: interfaces/builtin/upower_control.go [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+# Allow all access to UPower service
+dbus (receive, send)
+    bus=system
+    path=/org/freedesktop/UPower{,/**}
+    peer=(label=###SLOT_SECURITY_TAGS###),
+`
+
+const upowerControlConnectedPlugSecComp = `
+getsockname
```
Comment (jdstrand) [inline interfaces/builtin/upower_control.go]: getsockname is in the default seccomp template.

File: interfaces/builtin/upower_control.go [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/release"
+)
+
+const upowerControlPermanentSlotAppArmor = `
+# Description: Allow operating as the UPower service.
+
+network netlink,
+network netlink raw,
```
Comment (morphis) [inline interfaces/builtin/upower_control.go]: Yeah raw is needed. So let me remove 'network netlinkÄ

Comment (morphis) [inline interfaces/builtin/upower_control.go]: Yeah raw is needed. So let me remove 'network netlink'

File: interfaces/builtin/upower_control.go (hunk 2 above)
Comment (morphis) [inline interfaces/builtin/upower_control.go]: 'receive' is needed to get signals from the logind service.

File: interfaces/builtin/upower_control.go (hunk 3 above)
Comment (morphis) [inline interfaces/builtin/upower_control.go]: I will drop 'send' from the rule before and just leave this one with send. Then we have one to receive any signal from logind and one rule to allow calling just a specified set of methods. How does that sound?

File: interfaces/builtin/upower_control.go (hunk 6 above)
Comment (morphis) [inline interfaces/builtin/upower_control.go]: Done

**Comment by morphis** [comment]
Agreed with @jdstrand to merge both the upower-observe and the upower-control interfaces into a single upower-observe interface. Working on this now.

**Comment by morphis** [comment]
Closing as we merged all bits into the upower-observe interfaces on https://github.com/snapcore/snapd/pull/2469

**Pull request closed without merging**